specs/Print.md

```diff
@@ -0,0 +1,816 @@
+  // Return the user selected printer name.
+}
+
+// Function to get print settings for the selected printer.
+// You may also choose get the capabilties from the native printer API, display to the user to get
+// the print settings for the current web page and for the selected printer.
+PrintSettings AppWindow::GetSelectedPrinterPrintSettings(std::wstring printerName)
+{
+  PrintSettings printSettings;
+  printSettings.PrintBackgrounds = true;
+  printSettings.HeaderAndFooter = true;
+
+  return printSettings;
+
+  // or
+  //
+  // Use win32 DeviceCapabilitiesA function to get the capabilities of the selected printer.
+  // Display the printer capabilities to the user along with the page settings.
+  // Return the user selected settings.
+}
+
+bool AppWindow::PrintToPrinter()
+{
+  std::wstring printerName = GetPrinterName();
+  PrintSettings printInfo = GetSelectedPrinterPrintSettings(printerName);
+
+  wil::com_ptr<ICoreWebView2_15> webView2_15;
+  CHECK_FAILURE(m_webView->QueryInterface(IID_PPV_ARGS(&webView2_15)));
+
+  wil::com_ptr<ICoreWebView2Environment6> webviewEnvironment6;
+  CHECK_FAILURE(m_webViewEnvironment->QueryInterface(IID_PPV_ARGS(&webviewEnvironment6)));
+
+  wil::com_ptr<ICoreWebView2PrintSettings> printSettings = nullptr;
+  CHECK_FAILURE(webviewEnvironment6->CreatePrintSettings(&printSettings));
+
+  wil::com_ptr<ICoreWebView2PrintSettings2> printSettings2;
+  CHECK_FAILURE(printSettings->QueryInterface(IID_PPV_ARGS(&printSettings2)));
+
+  CHECK_FAILURE(printSettings2->put_Orientation(printInfo.Orientation));
+  CHECK_FAILURE(printSettings2->put_Copies(printInfo.Copies));
+  CHECK_FAILURE(printSettings2->put_PagesPerSheet(printInfo.PagesPerSheet));
+  CHECK_FAILURE(printSettings2->put_PageRanges(printInfo.Pages.c_str()));
+  if (printInfo.Media == COREWEBVIEW2_PRINT_MEDIA_SIZE_CUSTOM)
+  {
+    CHECK_FAILURE(printSettings2->put_PageWidth(printInfo.PaperWidth));
+    CHECK_FAILURE(printSettings2->put_PageHeight(printInfo.PaperHeight));
+  }
+  CHECK_FAILURE(printSettings2->put_ColorMode(printInfo.ColorMode));
+  CHECK_FAILURE(printSettings2->put_Collation(printInfo.Collation));
+  CHECK_FAILURE(printSettings2->put_Duplex(printInfo.Duplex));
+  CHECK_FAILURE(printSettings2->put_ScaleFactor(printInfo.ScaleFactor));
+  CHECK_FAILURE(printSettings2->put_ShouldPrintBackgrounds(printInfo.PrintBackgrounds));
+  CHECK_FAILURE(printSettings2->put_ShouldPrintHeaderAndFooter(printInfo.HeaderAndFooter));
+  CHECK_FAILURE(printSettings2->put_HeaderTitle(printInfo.HeaderTitle.c_str()));
+  CHECK_FAILURE(printSettings2->put_FooterUri(printInfo.FooterUri.c_str()));
+  CHECK_FAILURE(printSettings2->put_PrinterName(printerName.c_str()));
+
+  wil::unique_cotaskmem_string title;
+  CHECK_FAILURE(m_webView->get_DocumentTitle(&title));
+
+  CHECK_FAILURE(webView2_15->Print(
+      printSettings2.get(),
+      Callback<ICoreWebView2PrintCompletedHandler>(
+          [title = title.get(), this](HRESULT errorCode, BOOL isSuccessful) -> HRESULT
```

Comment:
If this handler is fired asynchronously from the Print method returning, `title` within the lambda now references a deleted pointer and `this` may not be valid anymore. Consider capturing by move and with a strong reference:

```c++
[title = std::move(title), this, self = com_ptr<AppWindow>(this)](...)
```

Reply:
Please take the `title` change for the sample code. We have a bug to figure out the right thing to do for `this` throughout the sample code.